### PR TITLE
fix(claude): honor CLAUDE_BIN_PATH in dev mode for libc-mismatch hosts

### LINKED ISSUE #1474: bug(claude): dev-mode workflows fail on glibc Linux due to SDK musl-first auto-resolution; honor CLAUDE_BIN_PATH as escape hatch
## Summary

  On glibc Linux hosts (Ubuntu/Debian/Fedora/etc.), `archon workflow run` fails immediately when running from source (dev mode). The Claude Code SDK auto-resolves its bundled binary in
  `[linux-x64-musl, linux-x64]` order, so the musl variant is selected first; its ELF interpreter (`/lib/ld-musl-x86_64.so.1`) doesn't exist on glibc systems, and the spawn fails with a
  misleading error that blames a missing path.

  The error currently surfaces as:

  ```
  ReferenceError: Claude Code native binary not found at
    .../node_modules/.bun/@anthropic-ai+claude-agent-sdk-linux-x64-musl@0.2.121
    /node_modules/@anthropic-ai/claude-agent-sdk-linux-x64-musl/claude
  ```

  The file is on disk and executable; the spawn failure is from the missing musl loader, not a missing binary.

  ## Reproduction

  1. Install Archon from source on a glibc host (any current Ubuntu/Debian/Fedora).
  2. `bun run dev:server` (or just `bun /home/$USER/.bun/bin/archon workflow run ...`).
  3. Run any workflow: `archon workflow run archon-assist --branch test "anything"`.
  4. Workflow fails in <1s with the error above.

  Both `@anthropic-ai/claude-agent-sdk-linux-x64-musl` and `@anthropic-ai/claude-agent-sdk-linux-x64` install via optionalDependencies on Linux (libc field in their `package.json` doesn't
  constrain Bun's install), and the SDK's resolver (function `N7` in `sdk.mjs`) tries musl first.

  Setting `CLAUDE_BIN_PATH=/home/<user>/.local/bin/claude` (the documented escape hatch) does **not** fix it: `binary-resolver.ts` early-returns `undefined` whenever
  `BUNDLED_IS_BINARY=false`, before the env var is even read. There is no working override for dev mode short of patching `node_modules`.

  ## Proposed fix

  Move the `CLAUDE_BIN_PATH` env-var check above the `BUNDLED_IS_BINARY` early return in `packages/providers/src/claude/binary-resolver.ts`. Config-file path stays binary-mode-only (it's
  per-repo, not per-machine — the env var is the right knob for libc-mismatch escape).

  Behavior preserved:
  - env var unset → identical to today (returns `undefined` in dev, falls through to autodetect/throw in binary mode)
  - env var set + file exists → resolved binary used (was already true in binary mode; now also true in dev mode)
  - env var set + file missing → clear error message (was already true in binary mode; now also true in dev mode)

  ## Patch

  ```diff
  diff --git a/packages/providers/src/claude/binary-resolver.ts b/packages/providers/src/claude/binary-resolver.ts
  index 6b918d44..0658374e 100644
  --- a/packages/providers/src/claude/binary-resolver.ts
  +++ b/packages/providers/src/claude/binary-resolver.ts
  @@ -64,9 +64,9 @@ const INSTALL_INSTRUCTIONS =
   export async function resolveClaudeBinaryPath(
     configClaudeBinaryPath?: string
   ): Promise<string | undefined> {
  -  if (!BUNDLED_IS_BINARY) return undefined;
  -
  -  // 1. Environment variable override
  +  // 1. Environment variable override — honored in dev mode too, so operators
  +  // on libc mismatches (e.g. glibc host with musl SDK variant first in the
  +  // resolution list) can pin a known-good binary without a compiled build.
     const envPath = process.env.CLAUDE_BIN_PATH;
     if (envPath) {
       if (!fileExists(envPath)) {
  @@ -80,6 +80,8 @@ export async function resolveClaudeBinaryPath(
       return envPath;
     }

  +  if (!BUNDLED_IS_BINARY) return undefined;
  +
     // 2. Config file override
     if (configClaudeBinaryPath) {
       if (!fileExists(configClaudeBinaryPath)) {
  ```

  ## Tests

  The existing `binary-resolver-dev.test.ts` asserts the old behavior (returns `undefined` even with env var set). It needs to be updated to reflect the new contract. Replacement:

  ```ts
  /**
   * Tests for the Claude binary resolver in dev mode (BUNDLED_IS_BINARY=false).
   * Separate file because binary-mode tests mock BUNDLED_IS_BINARY=true.
   *
   * Dev mode normally lets the SDK resolve the binary from its bundled
   * platform package. CLAUDE_BIN_PATH is honored as an escape hatch for
   * environments where SDK auto-resolution picks the wrong variant — most
   * notably glibc Linux hosts, where the SDK prefers the musl binary first
   * and silently falls over with a misleading "not found" error.
   * Config-file path is intentionally NOT honored in dev mode (still binary-only).
   */
  import { describe, test, expect, mock, beforeEach, afterAll, spyOn } from 'bun:test';
  import { createMockLogger } from '../test/mocks/logger';

  mock.module('@archon/paths', () => ({
    createLogger: mock(() => createMockLogger()),
    BUNDLED_IS_BINARY: false,
  }));

  import * as resolver from './binary-resolver';

  describe('resolveClaudeBinaryPath (dev mode)', () => {
    const originalEnv = process.env.CLAUDE_BIN_PATH;
    let fileExistsSpy: ReturnType<typeof spyOn>;

    beforeEach(() => {
      delete process.env.CLAUDE_BIN_PATH;
      fileExistsSpy?.mockRestore();
    });

    afterAll(() => {
      if (originalEnv !== undefined) {
        process.env.CLAUDE_BIN_PATH = originalEnv;
      } else {
        delete process.env.CLAUDE_BIN_PATH;
      }
      fileExistsSpy?.mockRestore();
    });

    test('returns undefined when nothing is configured', async () => {
      const result = await resolver.resolveClaudeBinaryPath();
      expect(result).toBeUndefined();
    });

    test('returns undefined when only config path is set (config is binary-mode only)', async () => {
      const result = await resolver.resolveClaudeBinaryPath('/some/custom/path');
      expect(result).toBeUndefined();
    });

    test('honors CLAUDE_BIN_PATH env var when file exists', async () => {
      process.env.CLAUDE_BIN_PATH = '/usr/local/bin/claude';
      fileExistsSpy = spyOn(resolver, 'fileExists').mockReturnValue(true);

      const result = await resolver.resolveClaudeBinaryPath();
      expect(result).toBe('/usr/local/bin/claude');
    });

    test('throws when CLAUDE_BIN_PATH is set but file does not exist', async () => {
      process.env.CLAUDE_BIN_PATH = '/nonexistent/claude';
      fileExistsSpy = spyOn(resolver, 'fileExists').mockReturnValue(false);

      await expect(resolver.resolveClaudeBinaryPath()).rejects.toThrow(
        'CLAUDE_BIN_PATH is set to "/nonexistent/claude" but the file does not exist'
      );
    });

    test('env var wins over config path in dev mode', async () => {
      process.env.CLAUDE_BIN_PATH = '/env/claude';
      fileExistsSpy = spyOn(resolver, 'fileExists').mockReturnValue(true);

      const result = await resolver.resolveClaudeBinaryPath('/config/claude');
      expect(result).toBe('/env/claude');
    });
  });
  ```

  Both test files pass locally:

  - `bun test packages/providers/src/claude/binary-resolver-dev.test.ts` → 5 pass
  - `bun test packages/providers/src/claude/binary-resolver.test.ts` → 9 pass (binary-mode tests, unaffected — env var was already step 1 there)

  ## Environment

  - OS: Ubuntu (glibc 2.33)
  - Bun: 1.3.13
  - Archon: dev branch at `bf1f471e`
  - Claude Code SDK: `@anthropic-ai/claude-agent-sdk@0.2.121`
  - Reproduces with `CLAUDECODE=1` set or unset (the nested-Claude warning is unrelated).

### PR DESCRIPTION
**Pull request by Wirasm**
## Summary

- **Problem**: On glibc Linux hosts (Ubuntu/Debian/Fedora), `archon workflow run` fails immediately in dev mode (`bun run dev:server`). The Claude Agent SDK auto-resolves its bundled native binary in `[linux-x64-musl, linux-x64]` order, picks the musl variant first, but the musl ELF interpreter (`/lib/ld-musl-x86_64.so.1`) does not exist on glibc — spawn fails and the SDK reports a misleading "binary not found" error.
- **Why it matters**: Affects every glibc Linux contributor running from source. The documented escape hatch `CLAUDE_BIN_PATH` is dead code in dev mode (early-returned before the env var is read), so the only workaround is patching `node_modules`.
- **What changed**: Hoisted the `CLAUDE_BIN_PATH` env-var block in `binary-resolver.ts` above the `BUNDLED_IS_BINARY` early return. Updated `binary-resolver-dev.test.ts` to reflect the new contract.
- **What did NOT change (scope boundary)**: Config-file path (`assistants.claude.claudeBinaryPath`) is intentionally still binary-mode-only — it's a per-repo setting, not a per-machine escape hatch. Autodetect and the throw-with-install-instructions paths also stay binary-mode-only. Binary-mode behavior is byte-identical to before.

### Before

```
glibc Linux user                Archon (dev mode)             Claude SDK
────────────────                ─────────────────             ──────────
bun run dev:server  ──────────▶ resolveClaudeBinaryPath()
                                  BUNDLED_IS_BINARY=false
                                  → returns undefined
                                pathToClaudeCodeExecutable    auto-resolves
                                  omitted                     [musl, glibc]
                                                              picks musl first
                                                              spawn fails (no
                                                              /lib/ld-musl-...)
                                                              ◀── ENOENT
sees "binary not found" ◀────── error surfaced verbatim

Setting CLAUDE_BIN_PATH=/home/me/.local/bin/claude does NOTHING:
  resolveClaudeBinaryPath() still returns undefined before reading env.
```

### After

```
glibc Linux user                Archon (dev mode)             Claude SDK
────────────────                ─────────────────             ──────────
export CLAUDE_BIN_PATH=         [reads env first]
  ~/.local/bin/claude
bun run dev:server  ──────────▶ resolveClaudeBinaryPath()
                                  *env set + file exists*
                                  → returns env path
                                pathToClaudeCodeExecutable=   spawn succeeds
                                  /home/me/.local/bin/claude  ──────────────▶
                                                              ◀── responds
sees response  ◀──────────────  streams to platform

env unset                  →  identical to today (returns undefined in dev)
env set + file missing     →  clear, actionable error pointing at the env path
```

### Before

```
┌────────────────────┐       ┌────────────────────────┐       ┌──────────────────┐
│ ClaudeProvider     │ ────▶ │ resolveClaudeBinaryPath│ ────▶ │ @anthropic-ai/   │
│ provider.ts        │       │ binary-resolver.ts     │       │ claude-agent-sdk │
└────────────────────┘       │   ┌──────────────────┐ │       │  (auto-resolve)  │
                             │   │ if (!BINARY)     │ │       └──────────────────┘
                             │   │   return undef ◀─┼─┼── env var unread in dev
                             │   │ env path?        │ │
                             │   │ config path?     │ │
                             │   │ autodetect?      │ │
                             │   └──────────────────┘ │
                             └────────────────────────┘
```

### After

```
┌────────────────────┐       ┌────────────────────────┐       ┌──────────────────┐
│ ClaudeProvider     │ ────▶ │ resolveClaudeBinaryPath│ ====▶ │ @anthropic-ai/   │
│ provider.ts        │       │ binary-resolver.ts [~] │       │ claude-agent-sdk │
└────────────────────┘       │   ┌──────────────────┐ │       │  (auto-resolve)  │
                             │   │ env path? [+]    │ │       └──────────────────┘
                             │   │ if (!BINARY)     │ │            ▲
                             │   │   return undef   │ │            │
                             │   │ config path?     │ │            │
                             │   │ autodetect?      │ │            │
                             │   └──────────────────┘ │            │
                             └────────────────────────┘            │
                                       ║                           │
                                       ╚═══ env path takes ════════╝
                                           precedence in both modes
```

**Connection inventory**:

| From | To | Status | Notes |
|------|----|--------|-------|
| ClaudeProvider | resolveClaudeBinaryPath | unchanged | Same call site |
| resolveClaudeBinaryPath | Claude SDK auto-resolve | **modified** | Now bypassed in dev when env var is set |
| `process.env.CLAUDE_BIN_PATH` | resolveClaudeBinaryPath | **modified** | Read in both modes (was binary-only) |

## Label Snapshot

- Risk: `risk: low`
- Size: `size: XS`
- Scope: `core` (closest available; lives in `@archon/providers`)
- Module: `providers:claude-binary-resolver`

## Change Metadata

- Change type: `bug`
- Primary scope: `core`

## Linked Issue

- Closes #1474
- Related #
- Depends on # (none)
- Supersedes # (none)

## Validation Evidence (required)

```bash
$ bun test packages/providers/src/claude/binary-resolver-dev.test.ts
 5 pass / 0 fail / 5 expect() calls (Ran 5 tests across 1 file)

$ bun test packages/providers/src/claude/binary-resolver.test.ts
 9 pass / 0 fail / 16 expect() calls (Ran 9 tests across 1 file)

$ bun run validate
exit 0  (check:bundled, type-check, lint, format:check, tests — all green)
```

- Evidence provided: All five `bun run validate` gates exit 0; both resolver test files pass (5 dev-mode + 9 binary-mode).
- If any command is intentionally skipped, explain why: None skipped.

## Security Impact (required)

- New permissions/capabilities? **No** (env-var override already existed in binary mode; this only enables it in dev mode)
- New external network calls? **No**
- Secrets/tokens handling changed? **No**
- File system access scope changed? **No** (still reads exactly the path the operator sets)
- If any `Yes`, describe risk and mitigation: N/A

## Compatibility / Migration

- Backward compatible? **Yes** (env unset → byte-identical to today; env set was previously ignored in dev, now honored — additive)
- Config/env changes? **No** (`CLAUDE_BIN_PATH` is documented; semantics broaden only)
- Database migration needed? **No**
- If yes, exact upgrade steps: N/A

## Human Verification (required)

- Verified scenarios:
  - Dev mode, env unset → resolver returns `undefined` (existing test)
  - Dev mode, env set + file exists → resolver returns env path (new test)
  - Dev mode, env set + file missing → resolver throws with actionable message naming the bad path (new test)
  - Dev mode, env set + config path also set → env wins (new test)
  - Dev mode, only config path set → config still ignored (new test, preserves old behavior)
  - Binary mode behavior unchanged (existing 9 tests still pass)
- Edge cases checked: env-var precedence over config in dev; no-op when env is unset (no surprise behavior change for current Mac/Windows users or CI).
- What was not verified: I did not reproduce the original musl-loader spawn failure locally (Mac host, glibc-only failure mode); the resolver-level fix is verified against the documented contract that the issue reporter validated end-to-end on Ubuntu glibc 2.33.

## Side Effects / Blast Radius (required)

- Affected subsystems/workflows: Only Claude provider binary resolution. Codex resolver is independent and untouched.
- Potential unintended effects: A user who has `CLAUDE_BIN_PATH` exported globally (e.g. for binary builds) and then runs dev mode will now have that path used instead of the SDK's bundled binary. This is the desired behavior (and matches what the env var name implies), but it is a behavior change for anyone who had the env var set without realizing dev mode was ignoring it. If the path is valid this is harmless; if the path is stale they'll see a clear error pointing at the path.
- Guardrails/monitoring for early detection: The error path includes the exact env-var value in its message ("CLAUDE_BIN_PATH is set to '<path>' but the file does not exist"), making misconfigurations self-diagnosing.

## Rollback Plan (required)

- Fast rollback command/path: `git revert bf8dabbe` — single-commit revert restores prior behavior.
- Feature flags or config toggles: None needed; users can simply `unset CLAUDE_BIN_PATH` to opt out without code changes.
- Observable failure symptoms: If the fix regresses, glibc dev users would see the original "musl binary not found" error again. Mac/Windows/binary users would see no behavior difference whatsoever.

## Risks and Mitigations

- Risk: A user with a stale `CLAUDE_BIN_PATH` set in their shell may now hit the resolver-level error in dev mode where they previously didn't.
  - Mitigation: The error message names the exact path that's bad and tells them how to fix it (verify path / unset env var). Lower-friction than the original musl-loader silent failure on glibc.
- Risk: Future SDK upgrades change the auto-resolution order in a way that makes the env-var workaround unnecessary.
  - Mitigation: This change is purely additive — when the env var is unset, behavior is byte-identical to today, so the workaround can stay forever without cost or be removed in a follow-up.

## Summary by CodeRabbit

* **Bug Fixes**
  * CLAUDE_BIN_PATH is now honored in dev mode; invalid paths produce clear errors and precedence rules are clarified.

* **Tests**
  * Expanded tests for env-var handling, precedence, and unset/invalid scenarios with reliable env setup/teardown.

* **Documentation**
  * Clarified dev vs binary-mode resolution and when to use the env var versus the config field.

* **Chores**
  * Updated changelog entry describing the fix.

### CODE REVIEW

**Comment by Wirasm** [comment]
## Review Summary

**Verdict**: minor-fixes-needed

Your fix correctly hoists `CLAUDE_BIN_PATH` above the `BUNDLED_IS_BINARY` early-return, enabling the escape hatch on Linux hosts where the SDK picks the wrong platform variant. The code quality, error handling, and test coverage are solid. What's blocking merge: three docs pages still describe dev mode as "auto-resolves via `node_modules` only" and don't mention that `CLAUDE_BIN_PATH` now works — users won't know the feature exists.

### Blocking issues
- **packages/docs-web/src/content/docs/getting-started/ai-assistants.md**: Add a note that `CLAUDE_BIN_PATH` can override the dev-mode auto-resolution. Example: "You can override with `CLAUDE_BIN_PATH` if the SDK picks the wrong platform binary (e.g. musl vs glibc on Linux)."
- **packages/docs-web/src/content/docs/getting-started/configuration.md**: Same update in the dev-mode configuration section.
- **packages/docs-web/src/content/docs/reference/troubleshooting.md**: Update "unaffected" language — `CLAUDE_BIN_PATH` can now override in dev mode, but config-file path remains binary-mode-only.

### Suggested fixes
- **CHANGELOG.md**: Add entry under `[Unreleased]` describing that `CLAUDE_BIN_PATH` is now honored in dev mode (previously ignored). Note that `assistants.claude.claudeBinaryPath` in config remains dev-mode-only.
- **binary-resolver.ts:70**: Add a test for empty-string `CLAUDE_BIN_PATH` (`''`) to verify it falls through to `undefined`.

### Minor / nice-to-have
- **binary-resolver-dev.test.ts:38**: For consistency with the binary-mode test file, drop `| undefined` from `fileExistsSpy` type and the `?` on `mockRestore()`.
- **binary-resolver-dev.test.ts:3-9** and **binary-resolver.ts:59-61**: The detailed glibc/musl failure description is valuable but risks drifting. Trim to a one-line "escape hatch for SDK binary resolution mismatches" framing.

### Compliments
- The PR body is thorough — architecture diagrams, actual test output, security/compatibility/rollback analysis, and human verification all included. Easy to review.
- The test `throws when CLAUDE_BIN_PATH is set but file does not exist` in dev mode directly covers the fail-fast improvement.
- The comment `"Config-file path is intentionally NOT honored in dev mode"` in the test file captures a policy decision that would otherwise confuse future maintainers.

---
*Reviewed via maintainer-review-pr workflow (Pi/Minimax). Aspects run: code-review, error-handling, test-coverage, comment-quality, docs-impact.*
